tests/xml_format_test.php

```diff
@@ -0,0 +1,139 @@
+    public function data_deserialise() {
+        return [
+            [
+                'body'   => '<request><some>text</some></request>',
+                'expect' => [
+                    'some' => 'text',
+                ],
+            ],
+            [
+                'body'   => '<request><some>1</some></request>',
+                'expect' => [
+                    'some' => 1,
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider data_deserialise
+     */
+    public function test_deserialise($body, $expect) {
+        $format = new xml_format();
+        $this->assertEquals($expect, $format->deserialise($body));
+    }
+
+
+    /**
+     * @expectedException invalid_parameter_exception
+     * @expectedExceptionMessage mangled and hideous though it was, request body could not be parsed as valid xml
+     */
+    public function test_deserialise_throws() {
+        $format = new xml_format();
+        $format->deserialise('<trololol');
+    }
+
+    public function test_send_error() {
+        $format = new xml_format();
+        $exception = new Exception('message', 1);
+
+        ob_start();
+        $format->send_error($exception);
+        $output = ob_get_contents();
+        ob_end_clean();
+
+        $this->assertRegExp('%\<response\>.*\</response\>%', $output);
+        $this->assertRegExp('%\<success\>false\</success\>%', $output);
+        $this->assertRegExp('%\<exception%', $output);
+        $this->assertRegExp('%class="Exception"%', $output);
+        $this->assertRegExp('%code="1"%', $output);
+        $this->assertRegExp('%\<message\>message\</message\>%', $output);
+    }
+
+    public function data_send_response() {
+        return [
+            [
+                'result' => true,
+                'description' => new external_value(PARAM_BOOL),
+                'expect' => [
+                    '%<value>true</value>%',
+                ],
+            ],
+            [
+                'result' => [
+                    'key' => 3.14,
+                ],
+                'description' => new external_single_structure([
+                    'key' => new external_value(PARAM_FLOAT),
+                ]),
+                'expect' => [
+                    '%<key>3.14</key>%',
+                ],
+            ],
+            [
+                'result' => [
+                    [
+                        'key' => 3.14,
+                    ],
+                ],
+                'description' => new external_multiple_structure(new external_single_structure([
+                    'key' => new external_value(PARAM_FLOAT),
+                ])),
+                'expect' => [
+                    '%<key>3.14</key>%',
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider data_send_response
+     */
+    public function test_send_response($result, external_description $description, $expect) {
```

Comment:
When running this test, I get this warning:

`1) Warning
The data provider specified for webservice_loadedrest_xml_format_testcase::test_send_response is invalid.
Class 'external_value' not found
`

Doesn't quite make sense, as `new external_value()` statements inside `data_send_response()` easily lead to `/lib/externallib.php` from inside PhpStorm.

Reply:
Can you paste the output of `$ vendor/bin/phpunit --group webservice_loadedrest` here? I'm interested in seeing the version information.

Reply:
Here it is:

`Moodle 3.4.2+ (Build: 20180405)
Php: 7.0.30.0.0.16.04.1, sqlsrv: 14.00.1000, OS: Linux 4.4.0-127-generic x86_64

PHPUnit 6.5.7 by Sebastian Bergmann and contributors.

......`

Reply:
Fixed via include of `externallib` -- it seems data providers execute prior to the PHPUnit bootstrap, which means the classloader isn't around to load the missing `external_*` classes.